packages/react-components/react-virtualizer/src/components/Virtualizer/Virtualizer.types.ts

```diff
@@ -0,0 +1,128 @@
+import type { ReactNode, MutableRefObject } from 'react';
+import type { ComponentProps, ComponentState, Slot } from '@fluentui/react-utilities';
+
+export type VirtualizerSlots = {
+  before: NonNullable<Slot<'div', 'td'>>;
+  beforeContainer: NonNullable<Slot<'div', 'tr'>>;
+  after: NonNullable<Slot<'div', 'td'>>;
+  afterContainer: NonNullable<Slot<'div', 'tr'>>;
+};
+
+// Virtualizer render function to procedurally generate children elements.
+// Q: Should we be using generic typing and passing through object data or a simple index system?
+export type VirtualizerChildRenderFunction = (index: number) => ReactNode;
+
+export type VirtualizerState = ComponentState<VirtualizerSlots> & {
+  /**
+   * The current virtualized array of children to show in the DOM.
+   */
+  virtualizedChildren: ReactNode[];
+  /**
+   * The current start index for the virtualizer, all previous index's will be removed from DOM.
+   */
+  virtualizerStartIndex: number;
+  /**
+   * Current buffer height required at beginning of array.
+   */
+  afterBufferHeight: number;
+  /**
+   * Current buffer height required at end of array.
+   */
+  beforeBufferHeight: number;
+  /**
+   * The total current height of the scrollView/child content.
+   */
+  totalVirtualizerHeight: number;
+  /**
+   * The scroll direction (vertical vs horizontal) - defaults to Vertical if not provided
+   */
+  horizontal: boolean;
+  /**
+   * Tells the virtualizer to measure in the reverse direction (for column-reverse order etc.)
+   */
+  reversed?: boolean;
+  /**
+   * Tells the virtualizer how much
+   */
+  bufferSize: number;
+};
+
+export type VirtualizerProps = ComponentProps<Partial<VirtualizerSlots>> & {
+  /**
+   * Required!
+   * Render function for rows
+   */
+  children: VirtualizerChildRenderFunction;
+
+  /**
+   * Required!
+   * Default row height to use if no custom callback provided.
+   * You must provide a value, even if it is just an assumed average or minimum size.
+   */
+  itemSize: number;
+
+  /**
+   * Required!
+   * The total number of items in the non-virtualized list.
+   */
+  numItems: number;
+
+  /**
+   * Number of elements to render in the list after virtualization.
+   * This needs to be enough items to cover the list's viewport
+   * as well as the buffer overhangs (bufferItems/bufferSize)
+   */
+  virtualizerLength: number;
```

Comment:
could you describe the relationship between `virtualizationLength` and `bufferItems`? from the current docstring I understand that they are the same concept but they are before/after the virtualized elements

Reply:
Added some better docstrings and constraints to showcase how they relate to each other here.